File: src/main/java/de/asmax/simpleperms/groups/GroupManager.java
package de.***REMOVED***.simpleperms.groups;

import de.***REMOVED***.simpleperms.Main;
import de.***REMOVED***.simpleperms.utils.Config;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.List;

public class GroupManager {
    //TODO: Add Players to groups and inherit their permissions
    private static Config config = Main.getInstance().getConfiguration();

    public static void addGroup(String group, int level) {
        ArrayList<String> tempList = new ArrayList<String>();
        ArrayList<String> playertempList = new ArrayList<String>();
        tempList.add("spacer.space");
        playertempList.add("DoNotTouchThis");

        config.getConfig().set("Groups." + group + ".level", level);
        config.getConfig().set("Groups." + group + ".permissions", tempList);
        config.getConfig().set("Groups." + group + ".players", playertempList);
        config.save();
    }

    public static void addPlayerToGroup(Player player, String group) {
        if(!getGroup(group)) {
            return;
        }

        if(player == null) {
            return;
        }

        String UUID = player.getUniqueId().toString();

        @SuppressWarnings("unchecked")
        List<String> playerList = (List<String>) config.getConfig().getList("Groups." + group + ".players");
        System.out.println(playerList);

        if (playerList == null) {
            System.out.println("PlayerList equals null");
            return;
        }
        playerList.add(UUID);
        config.getConfig().set("Groups." + group + ".players", playerList);
        config.save();
    }

    public static void removePlayerFromGroup(Player player, String group) {

        String UUID = player.getUniqueId().toString();

        @SuppressWarnings("unchecked")
        List<String> playerList = (List<String>) config.getConfig().getList("Groups." + group + ".players");

        if(!playerList.contains(UUID)) {
            return;
        }

        playerList.remove(UUID);
        config.getConfig().set("Groups." + group + ".players", playerList);
        config.save();
    }

    public static boolean getPlayerGroup(Player player, String group) {
        String UUID = player.getUniqueId().toString();

        @SuppressWarnings("unchecked")
        List<String> playerList = (List<String>) config.getConfig().getList("Groups." + group + ".players");

        if(playerList.contains(UUID)) {
            return true;
        } else {
            return false;
        }
    }

    public static void addPermission(String group, String permission) {
        System.out.println("[DEBUG] Group: " + group + " Permission: " + permission);
        @SuppressWarnings("unchecked")
        List<String> groupPermissionList = (List<String>) config.getConfig().getList("Groups." + group + ".permissions");

        if(config.getConfig().get("Groups." + group) == null) {
            return;
        }

        groupPermissionList.add(permission);
        config.getConfig().set("Groups." + group + ".permissions", groupPermissionList);
        config.save();

    }

    public static List<String> listPlayers(String group) {
        @SuppressWarnings("unchecked")
        List<String> playerList = (List<String>) config.getConfig().getList("Groups." + group + ".players");

        return playerList;
    }

    public static void copyDefaultGroup() {

        if(config.getConfig().get("Groups.default") != null) {
            return;
        }

        config.getConfig().set("Groups.default.level", 1);
        config.save();
    }

    public static void removePermission(String group, String permission) {
        List<String> groupPermissionList = (List<String>) config.getConfig().getList("Groups." + group + ".permissions");

        if(config.getConfig().get("Groups." + group) == null) {
            return;
        }

        if(!groupPermissionList.contains(permission)) {
            return;
        }

        groupPermissionList.remove(permission);
        config.getConfig().set("Groups." + group + ".permissions", groupPermissionList);
        config.save();
    }

    public static boolean getPermission(String group, String permission) {
        List<String> groupPermissionList = (List<String>) config.getConfig().getList("Groups." + group + ".permissions");

        if(config.getConfig().get("Groups." + group) == null) {
            return false;
        }

        if(groupPermissionList.contains(permission)) {
            return true;

        } else {
            return false;
        }

    }

    public static void removeGroup(String name) {

        if(config.getConfig().get("Groups." + name) == null) {
            return;
        }

        config.getConfig().set("Groups." + name, null);
        config.save();

    }

    public static String listGroups() {

        if(config.getConfig().get("Groups") == null) {
            return "error";
        }

        String groups = config.getConfig().getConfigurationSection("Groups").getKeys(false).toString();
        return groups;
    }

    public static boolean getGroup(String name) {

        if(config.getConfig().get("Groups." + name) != null) {
            return true;

        } else {

            return false;
        }
    }

    public static String listPermissions(String group) {

        if(config.getConfig().get("Groups." + group) == null) {
            return "error";
        }

        List<String> groupPermissionList = (List<String>) config.getConfig().getList("Groups." + group + ".permissions");

        String permissions = groupPermissionList.toString();
        return permissions;
    }

}

File: src/main/java/de/asmax/simpleperms/groups/GroupManager.kt
package de.***REMOVED***.simpleperms.groups

import de.***REMOVED***.simpleperms.Main
import de.***REMOVED***.simpleperms.utils.Config
import org.bukkit.entity.Player

object GroupManager {
    //TODO: Add Players to groups and inherit their permissions
    fun addGroup(group: String, level: Int) {
        val tempList = ArrayList<String>()
        val playertempList = ArrayList<String>()
        tempList.add("spacer.space")
        playertempList.add("DoNotTouchThis")
        Config.getConfig()["Groups.$group.level"] = level
        Config.getConfig()["Groups.$group.permissions"] = tempList
        Config.getConfig()["Groups.$group.players"] = playertempList
        Config.save()
    }

    fun addPlayerToGroup(player: Player?, group: String) {
        if (!getGroup(group)) {
            return
        }
        if (player == null) {
            return
        }
        val UUID = player.uniqueId.toString()
        val playerList = Config.getConfig().getList("Groups.$group.players") as MutableList<String>?
        println(playerList)
        if (playerList == null) {
            println("PlayerList equals null")
            return
        }
        playerList.add(UUID)
        Config.getConfig()["Groups.$group.players"] = playerList
        Config.save()
    }

    fun removePlayerFromGroup(player: Player, group: String) {
        val UUID = player.uniqueId.toString()
        val playerList = Config.getConfig().getList("Groups.$group.players") as MutableList<String>
        if (!playerList.contains(UUID)) {
            return
        }
        playerList.remove(UUID)
        Config.getConfig()["Groups.$group.players"] = playerList
        Config.save()
    }

    fun getPlayerGroup(player: Player, group: String): Boolean {
        val UUID = player.uniqueId.toString()
        val playerList = Config.getConfig().getList("Groups.$group.players") as List<String>
        return playerList.contains(UUID)
    }

    fun addPermission(group: String, permission: String) {
        println("[DEBUG] Group: $group Permission: $permission")
        val groupPermissionList = Config.getConfig().getList(
            "Groups.$group.permissions"
        ) as MutableList<String>
        if (Config.getConfig()["Groups.$group"] == null) {
            return
        }
        groupPermissionList.add(permission)
        Config.getConfig()["Groups.$group.permissions"] = groupPermissionList
        Config.save()
    }

    fun listPlayers(group: String): List<String> {
        return Config.getConfig().getList("Groups.$group.players") as List<String>
    }

    fun copyDefaultGroup() {
        if (Config.getConfig()["Groups.default"] != null) {
            return
        }
        Config.getConfig()["Groups.default.level"] = 1
        Config.save()
    }

    fun removePermission(group: String, permission: String) {
        val groupPermissionList = Config.getConfig().getList(
            "Groups.$group.permissions"
        ) as MutableList<String>
        if (Config.getConfig()["Groups.$group"] == null) {
            return
        }
        if (!groupPermissionList.contains(permission)) {
            return
        }
        groupPermissionList.remove(permission)
        Config.getConfig()["Groups.$group.permissions"] = groupPermissionList
        Config.save()
    }

    fun getPermission(group: String, permission: String): Boolean {
        val groupPermissionList = Config.getConfig().getList("Groups.$group.permissions") as List<String>
        if (Config.getConfig()["Groups.$group"] == null) {
            return false
        }
        return groupPermissionList.contains(permission)
    }

    fun removeGroup(name: String) {
        if (Config.getConfig()["Groups.$name"] == null) {
            return
        }
        Config.getConfig()["Groups.$name"] = null
        Config.save()
    }

    fun listGroups(): String {
        return if (Config.getConfig()["Groups"] == null) {
            "error"
        } else Config.getConfig().getConfigurationSection("Groups")?.getKeys(false).toString()
    }

    fun getGroup(name: String): Boolean {
        return Config.getConfig()["Groups.$name"] != null
    }

    fun listPermissions(group: String): String {
        if (Config.getConfig()["Groups.$group"] == null) {
            return "error"
        }
        val groupPermissionList = Config.getConfig()
            .getList("Groups.$group.permissions") as List<String>
        return groupPermissionList.toString()
    }
}